### PR TITLE
[황준영] sprint2

### PR DESCRIPTION
**Pull request by OfficialHwempire**
### 기본
- [x] 기본 항목 1
- [x] 기본 항목 2

### 심화
- [x] 심화 항목 1
- [x] 심화 항목 2

## 주요 변경사항
- Basic*Service, JCF*repository, File*repository 구현
- Basic*Service, JCF*repository, File*repository 테스트 구현 (Junit5 사용)

## 스크린샷
image

## 멘토에게
- 테스트 코드가 상당히 불안정합니다 
- 자주 코드를 갈아 엎어서 불필요한 커밋 로그가 많습니다.

### CODE REVIEW

File: .idea/.gitignore
Comment (ch4570) [inline .idea/.gitignore:1]: .idea 폴더는 IDE에서 사용하는 파일이기 때문에 프로젝트 최상위의 .gitignore에서 관리해서 제외하도록 하는게 좋을 것 같아요!

File: 7-sprint-mission/.idea/vcs.xml
Comment (ch4570) [inline 7-sprint-mission/.idea/vcs.xml:1]: .idea 폴더는 프로젝트에서 반드시 필요한 파일은 아니기 때문에 .gitignore 처리가 있으면 좋겠습니다!

File: src/main/java/com/sprint/mission/discodeit/deletedCash/DeletedChannel.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package com.sprint.mission.discodeit.deletedCash;
+
+import java.io.Serializable;
+
+public class DeletedChannel implements Serializable {
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/deletedCash/DeletedChannel.java:5]: 현업에서는 삭제된 데이터의 경우 일정 기간동안 보관해야 하는 정책 또는 법규가 존재할 수 있어요.
이런 경우에 데이터가 너무 많아지면 아카이빙 하는 별도의 데이터베이스를 가지고 있는 경우도 있지만, 대부분 삭제 여부를 필드로 관리하는 경우도 많이 있어요.

삭제된 데이터를 관리할 필요가 있다면 엔티티의 필드로 관리하는 게 더 좋을 것 같아요

File: src/main/java/com/sprint/mission/discodeit/dto/ChannelDto.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
+package com.sprint.mission.discodeit.dto;
+
+import com.sprint.mission.discodeit.entity.Entity;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+
+public class ChannelDto extends Entity{
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/dto/ChannelDto.java:10]: dto 클래스는 엔티티와 별도로 분리되어야 하는 클래스입니다.
Entity를 상속받게 되면 엔티티와 결합되기 때문에 만약 super class가 필요하다면 BaseDto와 같은 클래스를 따로 만드는 것이 좋을 것 같아요

File: src/main/java/com/sprint/mission/discodeit/entity/Channel.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+public class Channel extends Entity implements Serializable{
+    private String name;
+    private String description;
+    private boolean isPublic;
+    private boolean isTextChannel;
+    private final List<User> userDb;
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Channel.java:18]: User를 리스트로 가지고 있는 것은 매우 좋은 선택입니다!
채널에 어떤 유저가 참여하고 있는지 알 수 있기 때문이에요.

userDb 라는 이름은 유저 데이터 베이스로 오해할 수 있기 때문에 joinMemberList 또는 joinMembers 라는 이름이 더 좋지 않을까요??

File: src/main/java/com/sprint/mission/discodeit/entity/Channel.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        userDb.add(user);
+    }
+
+    public List<User> getUserDb() {
+        return userDb;
+    }
+
+    public void removeUserFromChannel(User user){
+
+        if(userDb.contains(user))
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Channel.java:30]: 조건문을 사용할때 브레이스를 너무 넓게 벌리면 코드 가독성에 안좋은 영향을 끼쳐요

```java
if (userDb.contains(user) {
   userDb.remove(user);
   return;
}
```

과 같이 코드를 작성하면 가독성이 더 증가할거에요

File: src/main/java/com/sprint/mission/discodeit/entity/Channel.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        {
+            userDb.remove(user);
+            return;
+        }
+
+            throw new IllegalArgumentException(StaticString.CHANNEL_NOT_EXIST);
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Channel.java:36]: 삭제 작업에 성공하면 return 문으로 탈출하는 것보다 만약 존재하지 않는 회원이 채널에서 탈퇴하려고 하면 예외를 발생시키는게 조금 더 자연스러운 동작일 것 같아요

File: src/main/java/com/sprint/mission/discodeit/entity/Channel.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        if (!(o instanceof Channel channel)) return false;
+        return isPublic == channel.isPublic && isTextChannel == channel.isTextChannel && Objects.equals(name, channel.name) && Objects.equals(description, channel.description);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, description, isPublic, isTextChannel);
+    }
+
+    public Channel(String name, String description, boolean isPublic, boolean isTextChannel) {
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Channel.java:51]: 매개변수가 다른 생성자가 여러개 필요하다면, 생성자를 계속 만드는 것보다는 가장 많은 매개변수를 가진 생성자를 두고, 생성자 오버로딩을 이용해보면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/entity/Entity.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package com.sprint.mission.discodeit.entity;
+
+import java.io.Serializable;
+import java.util.UUID;
+
+public class Entity implements Serializable {
+
+    private final UUID id;
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Entity.java:8]: id는 각 엔티티에서 관리하는게 조금 더 이상적일 수 있어요.
엔티티 별로 id의 타입이 다를 수 있기 때문이에요.

File: src/main/java/com/sprint/mission/discodeit/entity/Entity.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    private final long createdAt;
+    private long updatedAt;
+    public static final long DEFAULT_UPDATED_AT = 0L;
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Entity.java:11]: 엔티티가 처음 생성된 경우에는 현재 시각으로 값을 넣어주는게 더 자연스러울 거 같아요.
데이터의 변경은 수정 뿐만 아니라 없던 데이터가 생긴 것도 update 행위라고 볼 수 있을 거 같아요.

File: src/main/java/com/sprint/mission/discodeit/entity/Message.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
+package com.sprint.mission.discodeit.entity;
+
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+public class Message extends Entity{
+
+    private String content;
+    private User sender;
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Message.java:10]: 메시지를 전송한 User의 정보는 전체 Object를 포함 관계로 가지고 있는 것 보다는 user를 식별할 수 있는 가장 고유한 값만 가지고 있는게 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/entity/Entity.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package com.sprint.mission.discodeit.entity;
+
+import java.io.Serializable;
+import java.util.UUID;
+
+public class Entity implements Serializable {
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/entity/Entity.java:6]: Entity는 너무 범용적인 이름일 수 있어요.
모든 엔티티의 최고 조상이 된다는 뜻으로 보면 BaseEntity나 CommonEntity 등의 이름을 사용하면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/repository/ChannelRepository.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package com.sprint.mission.discodeit.repository;
+
+import com.sprint.mission.discodeit.dto.ChannelDto;
+import com.sprint.mission.discodeit.dto.DeletedChannelDto;
+import com.sprint.mission.discodeit.dto.UserDto;
+import com.sprint.mission.discodeit.entity.Channel;
+
+import java.util.UUID;
+
+public interface ChannelRepository {
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/repository/ChannelRepository.java:10]: Repository는 데이터베이스에서 엔티티에 대해서 CRUD 역할을 수행해요.
DTO를 사용하기 보다는 Entity를 직접 사용하는게 더 자연스러울 수 있어요

File: src/main/java/com/sprint/mission/discodeit/repository/ChannelRepository.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    public void saveChannel(ChannelDto channelDto);
+    public void deleteChannel(ChannelDto channelDto);
+    public <T>void updateChannel(ChannelDto channelDto, Channel.channelElement channelElement, T updatedContent);
+
+    public ChannelDto[] getAllChannel();
+    public ChannelDto[] getUpdatedChannel();
+    public DeletedChannelDto[] getDeletedChannel();
+    public void addUserToChannel(UserDto userDto, ChannelDto channelDto);
+    public void deleteUserFromChannel(UserDto userDto, ChannelDto channelDto);
+
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/repository/ChannelRepository.java:25]: 모든 Repository에서 사용할 범용 interface에 추상화를 진행하고 custom하게 생성되어야 하는 메서드만 따로 정의하면 ISP 원칙에도 부합하지 않을까요?

File: src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+public class FileChannelRepository implements ChannelRepository {
+
+    private final String DataRootPath = "C:\\Users\\황준영\\Java-codeit\\7-sprint-mission\\src\\main\\java\\com\\sprint\\mission\\discodeit\\data\\";
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java:21]: 절대 경로로 file path를 작성하게 되면 다른 사람이 프로젝트를 다운받으면 이 코드는 사용할 수 없는 코드가 되어버려요.

프로젝트 내에 데이터 파일을 가지게 하고 상대 경로를 이용해보면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+    private File channelRepositoryFile = new File(channelRepositoryDataPath);
+    private File deletedChannelRepositoryFile = new File(deletedChannelRepositoryDataPath);
+
+    public FileChannelRepository() {
+        repositoryCheck();
+        resetChannelRepository();
+    }
+    @Override
+    public ChannelDto getChannelById(UUID channelId) {
+        return loadAllChannel().stream().filter(x->x.getId().equals(channelId)).map(this::channelToChannelDto).findFirst().orElse(null);
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java:33]: 한 개의 채널을 검색하기 위해 모든 데이터를 스캔하는 건 시간 복잡도 측면에서 좋지 않아요.
처음에 Repository를 만들때 내부 Map을 가지고 ID 값 기반으로 찾아주면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+    public DeletedChannelDto[] getDeletedChannel() {
+        return loadAllDeletedChannel().stream().map(this::deletedChannelToDeletedChannelDto).toArray(DeletedChannelDto[]::new);
+    }
+
+    private void repositoryCheck(){
+        if(!channelRepositoryFile.exists()){
+            try{
+                channelRepositoryFile.createNewFile();
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java:137]: 예외가 발생했을때는 출력하는 것만으로는 처리가 부족할 수 있어요.
클라이언트가 어떤 문제가 발생했는지 전혀 알 수 없기 때문이에요.

이런 경우에는 어떤 문제가 발생헀는지 Custom Exception으로 wrapping해서 되던져 주면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+        }
+
+    }
+    @Override
+    public void saveChannel(ChannelDto channelDto){
+
+        List<Channel> channels = loadAllChannel();
+        channels.add(channelDtoToChannel(channelDto));
+        saveAllChannel(channels);
+        return;
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/repository/file/FileChannelRepository.java:286]: saveAllChannel() 메서드에서 처리가 끝났다면 return 문을 작성하지 않아도 메서드는 정상 종료될거에요!

File: src/main/java/com/sprint/mission/discodeit/service/util/StaticString.java [hunk 17]
```diff
@@ -0,0 +1,33 @@
+package com.sprint.mission.discodeit.service.util;
+
+public class StaticString {
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/service/util/StaticString.java:3]: 상수 값만 제공하는 경우에는 abstract class로 정의하거나 ENUM class로 정의해보면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/service/util/ValidateOperator.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+import com.sprint.mission.discodeit.dto.MessageDto;
+import com.sprint.mission.discodeit.dto.UserDto;
+import com.sprint.mission.discodeit.repository.ChannelRepository;
+import com.sprint.mission.discodeit.repository.MessageRepository;
+import com.sprint.mission.discodeit.repository.UserRepository;
+import com.sprint.mission.discodeit.service.ValidateService;
+
+import java.util.Arrays;
+
+public class ValidateOperator implements ValidateService {
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/service/util/ValidateOperator.java:13]: Validator에 모든 종류의 클래스에 대해 유효성 검증 메서드를 작성한다면 너무 큰 클래스가 될 수 있어요.
클래스 별로 Validator를 따로 두면 어떨까요?

File: src/main/java/com/sprint/mission/discodeit/test/JcfServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+
+        channelService.updateChannel(channel1Dto, Channel.channelElement.NAME, "JavaUpdate");
+        channelService.readUpdatedChannel();
+
+        channelService.inviteUserToChannel(user2Dto,channel2Dto);
+        channelService.readChannel(channel2Dto);
+        userService.deleteUser(user2Dto);
+        messageService.readMessage(m2Dto);
+
```
Comment (ch4570) [inline src/main/java/com/sprint/mission/discodeit/test/JcfServiceTest.java:65]: test를 작성하려는 의도로 클래스를 만드셨다면, 각 동작을 수행한 후 검증이 필요한데 현재 클래스는 데이터를 검증하고 있지 않아요.
검증하려고 하는 부분을 검증하는 코드를 넣어주면 좋을 것 같아요

File: src/test/java/com/sprint/mission/discodeit/repository/file/FileChannelRepositoryTest.java [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class FileChannelRepositoryTest {
+    private FileChannelRepository fileChannelRepository = new FileChannelRepository();
+    private ChannelDto channelDtoSample = new ChannelDto("Sample","SampleChannel",true,true);
+    private UserDto userDtoSample = new UserDto("SampleUser", "SampleNick","SampleEmail" , true );
+    @BeforeEach
+    void setUp() {
+        fileChannelRepository.saveChannel(channelDtoSample);
```
Comment (ch4570) [inline src/test/java/com/sprint/mission/discodeit/repository/file/FileChannelRepositoryTest.java:19]: 테스트 용 데이터를 미리 만들어주는 건 각 테스트의 격리를 어렵게 해요.
각 테스트 별로 고유한 데이터를 생성하는 것이 더 좋지 않을까요??

File: src/test/java/com/sprint/mission/discodeit/repository/jcf/JCFChannelRepositoryTest.java [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    @DisplayName("deleteUserFromChannel Test")
+    void deleteUserFromChannel() {
+        //given
+        jcfChannelRepository.saveChannel(channelDtoSample);
+        jcfChannelRepository.addUserToChannel(userDtoSample,channelDtoSample);
+        //when
+        jcfChannelRepository.deleteUserFromChannel(userDtoSample,channelDtoSample);
+        ChannelDto actualResult = jcfChannelRepository.getChannelById(channelDtoSample.getId());
+        //then
+        Assertions.assertThat(actualResult.getUserDtoList().size()).isEqualTo(0);
```
Comment (ch4570) [inline src/test/java/com/sprint/mission/discodeit/repository/jcf/JCFChannelRepositoryTest.java:93]: List를 검증할때는 size를 0으로 비교하지 않고 isEmpty()라는 메서드를 사용하면 조금 더 깨끗한 코드를 작성할 수 있어요!

File: src/test/java/com/sprint/mission/discodeit/repository/file/FileMessageRepositoryTest.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+import com.sprint.mission.discodeit.entity.User;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.UUID;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class FileMessageRepositoryTest {
```
Comment (ch4570) [inline src/test/java/com/sprint/mission/discodeit/repository/file/FileMessageRepositoryTest.java:15]: 클래스 별로 테스트를 작성하는 습관은 아주 좋은 습관이에요!
한 가지 아쉬운 점이 있다면 정상 케이스 보다 더 신경써야 하는게 예외 케이스에 대한 테스트에요.

현업에서도 edge-case에 대한 테스트가 부족해서 뼈아픈 실수들이 많이 발생하곤 해요.
예외가 발생하는 부분에 대해서도 꼼꼼히 테스트를 작성하시면 더 안정성있는 좋은 코드를 작성하는데 도움이 될거에요!

**Review by ch4570 — APPROVED** [review]
